projects/helper/treasury.js

```diff
@@ -1,44 +1,136 @@
+  matic: 'polygon',
+  frax: 'fraxtal'
+};
+
+function getLlamaChain(debankChain) {
+  return debankToLlamaChain[debankChain] || debankChain;
+}
 
 function treasuryExports(config) {
-  const chains = Object.keys(config)
-  const exportObj = {}
-  chains.forEach(chain => {
-    let { ownTokenOwners = [], ownTokens = [], owners = [], tokens = [], blacklistedTokens = [] } = config[chain]
-    const tvlConfig = { permitFailure: true, ...config[chain], }
+  const { isComplex, complexOwners = [], ...chains } = config;
+
+  const exportObj = {};
+
+  let fetchedComplexData = null;
+  let fetchPromise = null;
+
+  async function getComplexData() {
+    if (!fetchedComplexData) {
+      if (!fetchPromise) {
+        fetchPromise = Promise.all(
+          complexOwners.map(id =>
+            axios.get(API_URL_COMPLEX, {
+              params: { id, is_all: true },
+              headers: {
+                'accept': 'application/json',
+                'AccessKey': ACCESSKEY,
+              },
+            }).then(r => ({ id, tokens: r.data }))
+          )
+        );
+      }
+      fetchedComplexData = await fetchPromise;
+    }
+
+    return fetchedComplexData;
+  }
+
+  Object.keys(chains).forEach(chain => {
+    let { ownTokenOwners = [], ownTokens = [], owners = [], tokens = [], blacklistedTokens= [] } = config[chain];
+
+    tokens = tokens.filter(addr => !!addr);
+    ownTokens = ownTokens.filter(addr => !!addr);
+    blacklistedTokens = blacklistedTokens.filter(addr => !!addr);
+
+    const tvlConfig = { permitFailure: true, ...config[chain] };
+
+    if (!Array.isArray(tvlConfig.tokens)) tvlConfig.tokens = [];
+    if (!Array.isArray(tvlConfig.blacklistedTokens)) tvlConfig.blacklistedTokens = [];
+
     if (chain === 'solana') {
-      tvlConfig.solOwners = owners
+      tvlConfig.solOwners = owners;
     } else if (config[chain].fetchCoValentTokens !== false) {
       if (ankrChainMapping[chain]) {
-        tvlConfig.fetchCoValentTokens = true
-        const { tokenConfig } = config[chain]
-        if (!tokenConfig) {
-          tvlConfig.tokenConfig = { onlyWhitelisted: false, }
-        }
+        tvlConfig.fetchCoValentTokens = true;
+        if (!tvlConfig.tokenConfig) tvlConfig.tokenConfig = { onlyWhitelisted: false };
       } else if (defaultTokens[chain]) {
-        tvlConfig.tokens = [tokens, defaultTokens[chain]].flat()
+        tvlConfig.tokens = [tokens, defaultTokens[chain]].flat();
       }
     }
 
-    tvlConfig.blacklistedTokens = [...ownTokens, ...blacklistedTokens]
+    tvlConfig.blacklistedTokens = [...ownTokens, ...blacklistedTokens];
 
-    if (chain === 'arbitrum') {
-      tvlConfig.tokens = [...tokens, ARB]
-    }
-    exportObj[chain] = { tvl: sumTokensExport(tvlConfig) }
+    if (chain === 'arbitrum') tvlConfig.tokens = [...tokens, ARB];
+    if (!Array.isArray(tvlConfig.tokens)) tvlConfig.tokens = [];
+
+    exportObj[chain] = {
+      tvl: isComplex
+        ? async (timestamp, ethBlock, chainBlocks, { api }) => {
```

Comment:
can change tvl adapter signature to `(api) => `
timestamp is no longer the first parameter

Reply:
on second thought, use [mergeExports](https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/helper/utils.js#L112) if isComplex flag is set